network-ntt/src/client.rs

```diff
@@ -0,0 +1,297 @@
+use chain_core::property::{Block, HasHeader, Header, TransactionId};
+use futures::{sync::mpsc, sync::oneshot};
+use network_core::client::{self as core_client, block::BlockService};
+use protocol::{
+    network_transport::LightWeightConnectionId, protocol::BlockHeaders, protocol::GetBlockHeaders,
+    protocol::GetBlocks, Inbound, Message, Response,
+};
+use std::collections::HashMap;
+use std::marker::PhantomData;
+use std::net::SocketAddr;
+use tokio::net::TcpStream;
+use tokio::prelude::*;
+
+/// A handle that can be used in order for communication
+/// with the client thread.
+pub struct ClientHandle<T: Block, Tx> {
+    channel: mpsc::UnboundedSender<Request<T>>,
+    phantom: PhantomData<Tx>,
+}
+
+/// Connect to the remote client. Returns future that can
+/// be run on any executor.
+pub fn connect<B, Tx>(
+    sockaddr: SocketAddr,
+) -> impl Future<
+    Item = (impl Future<Item = (), Error = ()>, ClientHandle<B, Tx>),
```

Comment:
This future result type looks a bit loaded. Could `ClientHandle` implement the lifecycle future? If not, make the item type a structure that implements the connection future and has a named accessor method for the handle.

Reply:
I'm not sure, the problem here is the structure of the program that we want to support.
My idea was that we can spawn a future, after that we can keep a handle in order to run communication. The problem with communication is that we should be listening and replying to the messages even we don't send any request. So I decided to separate concerns and return future that can be spawned in any meaningful way and a communication handle.

Reply:
My main concern here is stylistic: the anonymous tuple that does not provide clues how to use it or what its members are. I think it should be a named struct and encapsulate both the future and the handle. If the future type is too cumbersome to spell out explicitly, it's OK to leave it as is.